Functions for NWIS continuous data

kathryn-willi opened the pull request:
Hey Matt!

Here's a PR containing two new functions:

`TADA_listNWIS()` returns a spatial summary of available USGS NWIS daily data based on site numbers, state codes, or a spatial area.

`TADA_getNWIS()` Retrieves and tidies USGS NWIS daily values for selected sites, states, or areas based on parameter codes and date range.

I also added some {testthat} function tests to make sure they work as expected, and I added a brief demo of their use in the Module2 vignette.

These two functions fulfill TADA's want of accessing the USGS's continuous data. It is quite simple as-is (it only pulls continuous daily data, for example), but I'm hopeful it will act as a launching off point for other functionality in the future if they choose. 

Because I still have other updates to the package, I did not build these into the package yet. (So, when testing, you will need to source the ContinuousDataFunctions.R file on top of load_all(), I think!)

If you could please review the code and also test that the code in the examples for each function, the module, and the testthat workflows run for you? 

After this PR, I will submit another one containing updates to currently-existing geospatial functions. 

Thanks!
Katie

kathryn-willi committed:
NWIS functions beginning

kathryn-willi committed:
continuous nwis data tests

kathryn-willi committed:
start to vignette

kathryn-willi committed:
tweaks

kathryn-willi committed:
editorial changes

kathryn-willi committed:
add PR

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#' sites_specific <- TADA_listNWIS(sites = site_nums)
+#'
+#' # Example 3: Query by state
+#' sites_state <- TADA_listNWIS(states = "CA")
```
suggestion: This example in particular makes me think that a progress bar ([example](https://github.com/USEPA/EPATADA/blob/97a369b4c04731b904f2591970e6c07d5f15eacb/R/DataDiscoveryRetrieval.R#L1136)) and silencing {dataRetrieval} messages ([example](https://github.com/USEPA/EPATADA/blob/97a369b4c04731b904f2591970e6c07d5f15eacb/R/DataDiscoveryRetrieval.R#L255-L257)) could be beneficial for the function. For me this prints a huge list of numbers to the console while a large download takes place.

kathryn-willi replied:
love that idea!

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#' List available USGS daily NWIS data
+#' 
+#' @description
+#' Retrieves available data from USGS National Water Information System (NWIS) based on 
```
```suggestion
#' Retrieves available metadata from USGS National Water Information System (NWIS) based on 
```

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#' @description
+#' Retrieves available data from USGS National Water Information System (NWIS) based on 
+#' different spatial queries: area of interest (AOI), specific sites, or state boundaries.
+#' Returns a spatial {sf} object containing site information and available parameters.
```
nitpick: I don't think that EPATADA uses the {} package notation so I've tried avoiding it in my work

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#' }
+#'
+TADA_listNWIS <- function(aoi_sf = "null", states = "null", sites = "null"){
+  if (!sum(purrr::map_lgl(list(aoi_sf, states[1], sites[1]), ~ is.null(.x) || (is.character(.x) && .x == "null"))) %in% c(2, 3)) {
```
```suggestion
# Confirm only a single argument has been provided
  if (!sum(purrr::map_lgl(list(aoi_sf, states[1], sites[1]), ~ is.null(.x) || (is.character(.x) && .x == "null"))) %in% c(2, 3)) {
```

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#'   Returns an empty sf object with the same structure if no data is found.
+#'
+#' @details Only one of the query arguments (`aoi_sf`, `states`, or `sites`) 
+#' should be provided. The function will stop if none or more than one are provided.
```
suggestion: You might also add a note about spatial area limits here.

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+    # Make sure returned USGS object is in the same CRS as what the user-supplied AOI is in:
+    if(sf::st_crs(aoi_sf) != sf::st_crs(gage_sites)){
+      
+      print(paste0("The `aoi_sf` is in CRS = ", sf::st_crs(aoi_sf)$epsg, ". Returning NWIS sites in the same CRS."))
```
thought: Wondering if `message()` would be better here than `print()`

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+        {
+          suppressMessages(dataRetrieval::whatNWISdata(bBox = bbox, service = "dv"))
+        },
+        error = function(e) {
+          stop("At least one of your user-supplied features in 'aoi_sf' is too large - all features must be less than 118,078 square miles (roughly the area of Nevada). For state queries, please use the argument `state` instead of `aoi_sf`.")
+        }
+      )
+    }
```
question: Is this the only reason an error would be returned?

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+    }
+    
+    gage_sites <- gage_sites %>%
+      sf::st_as_sf(coords = c('dec_long_va', 'dec_lat_va'), crs = 4269) # USGS supplies crs 4326 always
```
note: I'm not entirely sure I follow what's going on on this line specifically, just want to make sure that it's intentional since there's the CRS check on L150 below. Probably just me not getting it.

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+  } else if (any(unlist(sites) != "null")){
+    
+    # Check and split 'sites' into chunks if necessary
+    # {dataRetrieval} crashes if site list is too big:
```
```suggestion
    # {dataRetrieval} is limited by site list length:
```

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+                      "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV", "WI", "WY",
+                      "PR")
+    
+    if (!any(states %in% valid_states)) {
```
suggestion: Could throw `toupper()` in here to head off capitalization issues?

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#'
+#' @examples
+#' \dontrun{
+#' #' # Example 1: Query by area of interest
```
```suggestion
#' # Example 1: Query by area of interest
```

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#' }
+#' 
+TADA_getNWIS <- function(aoi_sf = "null", states = "null", sites = "null", parameter_codes, start_date, end_date){
+  
```
```suggestion
  # Confirm only a single argument has been provided
```

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -0,0 +1,430 @@
+#' sites_specific <- TADA_getNWIS(sites = c("11530500", "11532500"), parameter_codes = c("00060", "00010"), start_date = "2020-01-01", end_date = "2020-12-31")
+#'
+#' # Example 3: Query by states
+#' nwis_data <- TADA_getNWIS(states = c("RI", "CO"), parameter_codes = c("00060", "00010"), start_date = "2020-01-01", end_date = "2020-01-02")
```
note: Same suggestion as for the above function RE progress bar/and silencing messages. 

I also want to make sure that it's intentional that this example specifically returns no data for the selected time frame. (Edit: this might be related to length like I mention on L384 of vignettes/TADAModule2.Rmd below)

kathryn-willi replied:
Interesting, this should return a lot of data (over 600 observations), as should the other example you've highlighted below. I assume both of these are server issues. This definitely highlights I need to do some more error handling as you suggested for when the server craps out!

mbrousil commented on vignettes/TADAModule2.Rmd:
````diff
@@ -359,3 +359,121 @@ TADA_ViewATTAINS(.data = TADA_with_ATTAINS_filled)
+Let's retrieve information about available NWIS data by a state, then visualize
+that information:
+
+```{r, eval=FALSE}
````
issue: The RI example doesn't return any data for me. Running `TADA_listNWIS(states = "RI")` I get `No daily USGS-NWIS data in selected query.` However, I'm wondering if this is a length issue rather than a data availability issue? 
When I run the following code:
```r
sites <-  dataRetrieval::whatNWISsites(stateCd = "RI")
sites <- sites %>% dplyr::bind_rows() %>% dplyr::distinct() %>% .$site_no
site_chunks <- if (length(sites) > 35000) {
  split(sites, ceiling(seq_along(sites) / 35000))
} else {
  list(sites)
}
dataRetrieval::whatNWISdata(siteNumber = site_chunks, service = "dv")

```
I get 
```
Error in `httr2::req_perform()`:
! HTTP 414 URI Too Long.
```
---
For testing I replaced the "RI" with `sites = c("11530500", "11532500")` and this section worked well for me, though!

mbrousil commented on vignettes/TADAModule2.Rmd:
```diff
@@ -359,3 +359,121 @@ TADA_ViewATTAINS(.data = TADA_with_ATTAINS_filled)
+      ri_summary$site_no,
+      "<br> Parameters: ",
+      ri_summary$parameters,
+      "<br> Parameter codess: ",
```
```suggestion
      "<br> Parameter codes: ",
```

mbrousil commented on vignettes/TADAModule2.Rmd:
```diff
@@ -359,3 +359,121 @@ TADA_ViewATTAINS(.data = TADA_with_ATTAINS_filled)
+cherokee_sf <- sf::st_read("inst/extdata/OKTribe.shp") %>%
+  dplyr::filter(NAME == "Cherokee")
+
+cherokee_nwis <- TADA_listNWIS(aoi_sf = cherokee_sf)
```
note: At times this line has failed for me (and L459 below), I assume because of quirks server-side. So I think that strengthens the case for changing the error message on L135 of R/ContinuousDataFunctions.R

kathryn-willi replied:
the bbox argument in dataRetrieval calls changed in the latest update; I think this is why a lot of the sf-based calls were busted on your end!

mbrousil reviewed (changes requested):
Hey KW,
This largely looks good but I did run into a few problems, which I've pointed out. I think they mostly relate to handling errors and inconsistencies that we get from the NWIS side of things. The tests all ran correctly for me though! Let me know if you have any questions.

kathryn-willi committed:
Update R/ContinuousDataFunctions.R

Co-authored-by: Matt Brousil <37380883+mbrousil@users.noreply.github.com>

kathryn-willi committed:
Update R/ContinuousDataFunctions.R

Co-authored-by: Matt Brousil <37380883+mbrousil@users.noreply.github.com>

kathryn-willi committed:
Update R/ContinuousDataFunctions.R

Co-authored-by: Matt Brousil <37380883+mbrousil@users.noreply.github.com>

kathryn-willi committed:
Update R/ContinuousDataFunctions.R

Co-authored-by: Matt Brousil <37380883+mbrousil@users.noreply.github.com>

kathryn-willi committed:
Update R/ContinuousDataFunctions.R

Co-authored-by: Matt Brousil <37380883+mbrousil@users.noreply.github.com>

kathryn-willi committed:
Update vignettes/TADAModule2.Rmd

Co-authored-by: Matt Brousil <37380883+mbrousil@users.noreply.github.com>

kathryn-willi committed:
overhaul

kathryn-willi commented:
@mbrousil - as mentioned on teams, I was using an old package version of {dataRetrieval}, and many of the issues you encountered were related to that. I've updated the functions accordingly, and I've also added a lot more error handling and messages. Can you try to run the functions again now that they've been modified? (Including the function examples, the vignette, and the testthat's...) I know this is annoying; sorry! 😭

PS I incorporated an update bar, but it wasn't really working super well with how the data is getting pulled and so I subsequently removed it. Also, with some of the function updates I had to make due to the package update, things are moving a lot faster than they were originally so I don't think it's as necessary/useful as it could have been before.

kathryn-willi committed:
text changes

mbrousil commented:
Awesome! These changes seem to have taken care of the issues I was having. The only thing that came up for me was with the RI example in the vignette. I got this warning: `Warning message: sf layer has inconsistent datum (+proj=longlat +datum=NAD83 +no_defs). Need '+proj=longlat +datum=WGS84'`.

mbrousil commented on R/ContinuousDataFunctions.R:
```diff
@@ -116,47 +120,91 @@ TADA_listNWIS <- function(aoi_sf = "null", states = "null", sites = "null"){
+    validate_aoi_size <- function(aoi_sf) {
+      # Get square mile conversion factor for the projection
+      # 2.58999e+6 converts square meters to square miles
+      sq_m_to_sq_miles <- 1/2.58999e+6
```
```suggestion
      sq_m_to_sq_miles <- 1 / 2.58999e+6
```

kathryn-willi committed:
Update R/ContinuousDataFunctions.R

Co-authored-by: Matt Brousil <37380883+mbrousil@users.noreply.github.com>

kathryn-willi commented:
Thanks for the quick turnaround on this, MB! Stay tuned for another TADA PR sometime next week. :)
